tools/deploy_firmware.py

```diff
@@ -0,0 +1,288 @@
+#!/usr/bin/env python3
+
+import time
+import sys
+import os
+import subprocess
+import difflib
+import argparse
+import shutil
+from typing import List
+
+import pexpect
+import pexpect.fdpexpect
+import serial
+
+
+class PrplwrtDevice:
+    """Represents a prplWrt device.
+
+        Offers methods to check if a device needs to be upgraded and to do the actual upgrade.
+
+        It needs to have access to the artifacts of a build job to determine when an upgrade
+        is needed (see `artifacts_dir`).
+    """
+
+    shell_prompt = "_pexpect_prompt_ "
+    """We will temporarily use this prompt on the device to make it easier to "expect" it."""
+
+    baudrate = 115200
+    """The baudrate of the serial connection to the device."""
+
+    def __init__(self, device: str, name: str, image: str, username: str = "root"):
+        """
+
+        Parameters
+        -----------
+        device: str
+            The name of the platform (example: netgear-rax40).
+        name: str
+            The name of the device (it should ne reachable through ssh without a password).
+        image: str
+            The name of the image that can be used to upgrade the device.
+        username: str, optional
+            The username to use when connecting to the device over SSH.
+        """
+        self.device = device
+        self.name = name
+        self.image = image
+        self.username = username
+
+        self.scriptdir = os.path.join(os.path.dirname(os.path.realpath(__file__)), "..")
+        self.rootdir = os.path.realpath(os.path.join(self.scriptdir))
+        self.artifacts_dir = os.path.join(self.rootdir, "build/{}".format(self.device))
+        """The directory where artifacts are stored. It's expected to contain
+        prplwrt-version, and the image file."""
+
+    def set_prompt(self, pexp: pexpect):
+        """Set the prompt to `shell_prompt` on a pexpect object."""
+        pexp.sendline("export PS1='{}'".format(self.shell_prompt))
+        pexp.sendline("")
+
+    def read_remote_version(self) -> List[str]:
+        """Read prplwrt-version on a remote device
+
+        Returns
+        -------
+        List[str]
+            The content of the prplwrt-version file on the device as a list of strings.
+        """
+        with pexpect.spawn("ssh {}@{}".format(self.username, self.name)) as shell:
```

Comment:
Because I initially thought I had to "manually" configure the location of the config file, as well as the ssh keys to use.
Apparently it's not needed, so I switched to pxssh.